### ISSUE TITLE
rancher2_app_v2 throws error about inconsistent results

### ISSUE THREAD

**Issue by MG2R** [rancher/terraform-provider-rancher2#540]
When trying to add a new `rancher2_app_v2` resource for Dynatrace, I got the following error on apply:

```
04-Dec-2020 14:44:17 | Error: Provider produced inconsistent result after apply
-- | --
04-Dec-2020 14:44:17 |  
04-Dec-2020 14:44:17 | When applying changes to
04-Dec-2020 14:44:17 | module.rancher_cluster["glent-dev"].module.rancher_cluster_configuration[0].rancher2_app_v2.dynatrace[0],
04-Dec-2020 14:44:17 | provider "registry.terraform.io/rancher/rancher2" produced an unexpected new
04-Dec-2020 14:44:17 | value: Root resource was present, but now absent.
04-Dec-2020 14:44:17 |  
04-Dec-2020 14:44:17 | This is a bug in the provider, which should be reported in the provider's own
04-Dec-2020 14:44:17 | issue tracker.
04-Dec-2020 14:44:17 |  
04-Dec-2020 14:44:17 | make: *** [apply-plan] Error 1
```

I got this by running the following plan:

```
04-Dec-2020 14:43:20 | Terraform will perform the following actions:
-- | --
04-Dec-2020 14:43:20 |  
04-Dec-2020 14:43:20 | # module.rancher_cluster["glent-dev"].module.rancher_cluster_configuration[0].rancher2_app_v2.dynatrace[0] will be created
04-Dec-2020 14:43:20 | + resource "rancher2_app_v2" "dynatrace" {
04-Dec-2020 14:43:20 | + annotations                 = (known after apply)
04-Dec-2020 14:43:20 | + chart_name                  = "dynatrace-oneagent-operator"
04-Dec-2020 14:43:20 | + chart_version               = "0.8.000"
04-Dec-2020 14:43:20 | + cleanup_on_fail             = false
04-Dec-2020 14:43:20 | + cluster_id                  = "c-krmhh"
04-Dec-2020 14:43:20 | + cluster_name                = (known after apply)
04-Dec-2020 14:43:20 | + disable_hooks               = false
04-Dec-2020 14:43:20 | + disable_open_api_validation = false
04-Dec-2020 14:43:20 | + force_upgrade               = false
04-Dec-2020 14:43:20 | + id                          = (known after apply)
04-Dec-2020 14:43:20 | + labels                      = (known after apply)
04-Dec-2020 14:43:20 | + name                        = "dynatrace"
04-Dec-2020 14:43:20 | + namespace                   = (known after apply)
04-Dec-2020 14:43:20 | + project_id                  = "c-krmhh:p-bhd7q"
04-Dec-2020 14:43:20 | + repo_name                   = "rancher-partner-charts"
04-Dec-2020 14:43:20 | + values                      = <<~EOT
04-Dec-2020 14:43:20 | ---
04-Dec-2020 14:43:20 | oneagent:
04-Dec-2020 14:43:20 | apiUrl: "REDACTED"
04-Dec-2020 14:43:20 | secret:
04-Dec-2020 14:43:20 | apiToken: "REDACTED"
04-Dec-2020 14:43:20 | paasToken: "REDACTED"
04-Dec-2020 14:43:20 | EOT
04-Dec-2020 14:43:20 | + wait                        = false
04-Dec-2020 14:43:20 | }
04-Dec-2020 14:43:20 |  
04-Dec-2020 14:43:20 | # module.rancher_cluster["glent-dev"].module.rancher_cluster_configuration[0].rancher2_namespace.dynatrace will be created
04-Dec-2020 14:43:20 | + resource "rancher2_namespace" "dynatrace" {
04-Dec-2020 14:43:20 | + annotations      = (known after apply)
04-Dec-2020 14:43:20 | + id               = (known after apply)
04-Dec-2020 14:43:20 | + labels           = (known after apply)
04-Dec-2020 14:43:20 | + name             = "dynatrace"
04-Dec-2020 14:43:20 | + project_id       = "c-krmhh:p-bhd7q"
04-Dec-2020 14:43:20 | + wait_for_cluster = false
04-Dec-2020 14:43:20 |  
04-Dec-2020 14:43:20 | + resource_quota {
04-Dec-2020 14:43:20 | + limit {
04-Dec-2020 14:43:20 | + config_maps              = (known after apply)
04-Dec-2020 14:43:20 | + limits_cpu               = (known after apply)
04-Dec-2020 14:43:20 | + limits_memory            = (known after apply)
04-Dec-2020 14:43:20 | + persistent_volume_claims = (known after apply)
04-Dec-2020 14:43:20 | + pods                     = (known after apply)
04-Dec-2020 14:43:20 | + replication_controllers  = (known after apply)
04-Dec-2020 14:43:20 | + requests_cpu             = (known after apply)
04-Dec-2020 14:43:20 | + requests_memory          = (known after apply)
04-Dec-2020 14:43:20 | + requests_storage         = (known after apply)
04-Dec-2020 14:43:20 | + secrets                  = (known after apply)
04-Dec-2020 14:43:20 | + services                 = (known after apply)
04-Dec-2020 14:43:20 | + services_load_balancers  = (known after apply)
04-Dec-2020 14:43:20 | + services_node_ports      = (known after apply)
04-Dec-2020 14:43:20 | }
04-Dec-2020 14:43:20 | }
04-Dec-2020 14:43:20 | }
04-Dec-2020 14:43:20 |  
04-Dec-2020 14:43:20 | Plan: 2 to add, 0 to change, 0 to destroy.
04-Dec-2020 14:43:20 |  
```

The plan was the result of the following code:

```
resource "rancher2_namespace" "dynatrace" {
  name       = "dynatrace"
  project_id = module.plumbing_workspace.project_id
}

resource "rancher2_app_v2" "dynatrace" {
  count = var.dynatrace_enable ? 1 : 0

  name       = "dynatrace"
  cluster_id = var.cluster_id

  repo_name     = "rancher-partner-charts"
  chart_name    = "dynatrace-oneagent-operator"
  chart_version = "0.8.000"

  project_id = module.plumbing_workspace.project_id
  namespace  = rancher2_namespace.dynatrace.id

  values = templatefile(
    "${path.module}/dynatrace/config.yml.tpl",
    var.dynatrace_settings
  )
}
```

The app was created successfully, although not all resources were rolled out successfully. When rerunning the plan/apply, it tries to create a new app, even though the old one is still there.

**Comment by mholasek** [comment]
Same problem here, trying to install longhorn via app_v2.
```terraform
resource "rancher2_app_v2" "longhorn" {  
  cluster_id = data.rancher2_cluster.local.id
  name = "rancher-longhorn"
  namespace = "longhorn-system"
  repo_name = "rancher-charts"
  chart_name = "longhorn"
  chart_version = "1.0.201"
  # values = file("values.yaml")
  count = var.enable_longhorn == true ? 1 : 0
}
```

Results in:
```bash
Error: Provider produced inconsistent result after apply

When applying changes to module.rancher2_apps.rancher2_app_v2.longhorn[0],
provider "registry.terraform.io/rancher/rancher2" produced an unexpected new
value: Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

**Comment by rawmind0** [comment]
When terraform shows errors like `...produced an unexpected new value..` ` ...produced inconsistent plan..`, most of the times use to be related to modules sync or vars sync. I guess the issue may be related with `count = var.* ? 1 : 0`. Terraform seems updating resource data in the meanwhile is applying. Error is saying that rancher2_app_v2.* tf definition was there to apply, but it isn't anymore when applied. Have you tested without this var??

**Comment by mholasek** [comment]
@rawmind0: I removed the count variable and unfortunately I get the same error.

**Comment by rawmind0** [comment]
@mholasek , you are right count argument is not causing the issue. I was able to reproduce it. There is a race condition if chart has dependencies and `wait = false` (default value).  Try adding `wait = true` to you app v2 longhorn definition, it should work fine.

Included a provider patch at PR #538 to set `wait = true` by default.

**Comment by MG2R** [comment]
Just tried this: adding `wait = true` does not fix it for me. I'm guessing the specific issue I'm running up against is that Rancher is unable to roll out all the chart resources (due to [this issue](https://github.com/Dynatrace/helm-charts/issues/57) in the Dynatrace chart). However, that shouldn't stop Terraform from handling this gracefully. The app was created in the Rancher, so it should be in the Terraform statefile. Right now it ends up in limbo for me.

**Comment by MG2R** [comment]
Another discrepancy I just noticed: the name of the installed app does not match the name I entered in Terraform. As you can see in the code in the original message, I set the name as `name = dynatrace`. However, after running Terraform (and it failing), the following app is installed in cluster explorer:

![dynatrace_app_debug_hpc5150](https://user-images.githubusercontent.com/5439592/101788434-61190300-3b00-11eb-8e30-31f137a7218c.PNG)

As you can see, the name does not match what was entered in Terraform.

I can't figure out why this is. I'm using `rancher2_app_v2` also to deploy two instances of filebeat, and there the name is correctly applied to the applications. (See screenshot)

![hpc5150](https://user-images.githubusercontent.com/5439592/101789496-a8ec5a00-3b01-11eb-84b7-2930bb1a5e00.PNG)

When I deploy the dynatrace-oneagent-operator through the UI, the name I give it there IS actually applied to the application.

**Comment by rawmind0** [comment]
@MG2R , you are right, i've reproduced the issue. The issue you are having is distinct from @mholasek, this is caused by an app v2 name bug at the provider. If you set `name = "dynatrace-oneagent-operator"` your app v2 should be deployed and managed properly by terraform.

The provider should force release name and namespace if chart has defined annotations `catalog.cattle.io/release-name` and/or `catalog.cattle.io/namespace` and if it's a certified rancher chart. This is not working properly, working on a fix

**Comment by mholasek** [comment]
> @mholasek , you are right count argument is not causing the issue. I was able to reproduce it. There is a race condition if chart has dependencies and `wait = false` (default value). Try adding `wait = true` to you app v2 longhorn definition, it should work fine.
> 
> Included a provider patch at PR #538 to set `wait = true` by default.


I have added wait = true. Unfortunately this does not help. The error still occurs. I install sequentially the monitoring chart and the longhorn chart. With the longhorn chart this error occurs.

**Comment by rawmind0** [comment]
@mholasek , i can reproduce your longhorn issue with `wait=false` but working fine if set `wait=true`
This is working fine for me, using current provider release and Rancher v2.5.2
```
resource "rancher2_app_v2" "longhorn" {  
  cluster_id = "CLUSTER_ID"
  name = "rancher-longhorn"
  namespace = "longhorn-system"
  repo_name = "rancher-charts"
  chart_name = "longhorn"
  chart_version = "1.0.201"
  wait = true
}
```

**Comment by rawmind0** [comment]
@MG2R , added a patch to fix `dynatrace` issue at PR #538

**Comment by MG2R** [comment]
Thank you! Just confirmed setting the name as `dynatrace-oneagent-operator` worked!
